src/python/pants/backend/python/mixed_interpreter_constraints/py_constraints.py

```diff
@@ -1,35 +1,152 @@
-from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.goal import Goal, GoalSubsystem, Outputting
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import (
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+    UnexpandedTargets,
+)
 from pants.python.python_setup import PythonSetup
 
 logger = logging.getLogger(__name__)
 
 
-class PyConstraintsSubsystem(GoalSubsystem):
+class PyConstraintsSubsystem(Outputting, GoalSubsystem):
     """Determine what Python interpreter constraints are used by files/targets."""
 
     name = "py-constraints"
 
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--summary",
+            type=bool,
+            default=False,
+            help=(
+                "Output a CSV with each per-target information about interpreter constraints. The "
+                "headers are `Target`, `Constraints`, `Transitive Constraints`, `# Dependencies`, "
+                "and `# Dependees`.\n\nThis information can be useful when prioritizing a Python "
+                "migration, such as Python 3 migrations. Use `# Dependencies` and `# Dependees` to "
+                "help prioritize which targets are easiest to port (low # dependencies) and "
+                "highest impact to port (high # dependees).\n\nYou may want to import into Excel "
+                "or Google Sheets to post-process/filter the data."
+            ),
+        )
+
+    @property
+    def summary(self) -> bool:
+        return cast(bool, self.options.summary)
+
 
 class PyConstraintsGoal(Goal):
     subsystem_cls = PyConstraintsSubsystem
 
 
 @goal_rule
 async def py_constraints(
-    addresses: Addresses, console: Console, python_setup: PythonSetup
+    addresses: Addresses,
+    console: Console,
+    py_constraints_subsystem: PyConstraintsSubsystem,
+    python_setup: PythonSetup,
 ) -> PyConstraintsGoal:
+    if py_constraints_subsystem.summary:
+        if addresses:
+            console.print_stderr(
+                "The `py-constraints --summary` goal does not take file/target arguments. It will "
+                "generate a summary of your entire repository. You can then use Google "
+                "Sheets/Excel to post-process/filter the data."
+            )
+            return PyConstraintsGoal(exit_code=1)
+
+        all_expanded_targets, all_explicit_targets = await MultiGet(
+            Get(Targets, AddressSpecs([DescendantAddresses("")])),
+            Get(UnexpandedTargets, AddressSpecs([DescendantAddresses("")])),
+        )
+        all_python_targets = sorted(
+            {
+                t
+                for t in (*all_expanded_targets, *all_explicit_targets)
+                if t.has_field(InterpreterConstraintsField)
+                or t.has_field(PythonInterpreterCompatibility)
```

Comment:
Is this filtering something that could/should be accomplished with a FieldSet?

Reply:
It could be, but shouldn't be. FieldSets are much more complex than using the Target API directly. They really only exist to solve the problem of needing to pass specific types to the engine. Unless you're using a union, there's no need for field sets.